<a href="https://colab.research.google.com/github/Tufayel5/Assignment-3/blob/main/Assignment3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Imports section
import pandas as pd
import numpy as np

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
temp_data = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv") 

In [3]:
# Output the first 15 rows of the data
temp_data.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [24]:
# Display a summary of the table information (number of datapoints, etc.)
print(temp_data.describe())
print()
print("It has " , temp_data.size , " number of datapoints.")

       Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

It has  3000  number of datapoints.


## Part 2. Splitting the dataset

In [7]:
# Take the pandas dataset and split it into our features (X) and label (y)
features = temp_data[["Temperature °C", "Mols KCL"]]
label = temp_data["Size nm^3"] 

In [8]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.1, train_size=0.9) 

## Part 3. Perform a Linear Regression

In [25]:
# Use sklearn to train a model on the training set
from sklearn.linear_model import LinearRegression

LinRegression = LinearRegression().fit(X_train, y_train) # we use train for fitting
LinRegression.score(X_test, y_test) # we use test for scoring

0.8501254996725837

In [26]:
# Create a sample datapoint and predict the output of that sample with the trained model
print("Prediction: ",LinReg.predict(np.array([[1,2]]))) # we create a 2D array for features
print("Coefficients: ",LinRegression.coef_) # we will get 2 coefficients from this
print("Intercept: ", LinRegression.intercept_)

Prediction:  [-410835.85193939]
Coefficients:  [ 866.66028602 1038.82268793]
Intercept:  -413780.1576012807


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


In [16]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

$S = -413500.37192 +  845.72429(a) + 1020.15832(b) $
<p>My equation derives from the regression of intercept plus the regression of coefficients, a and b which fits the prediction. This model does not reach 100% accuracy. </p>




## Part 4. Use Cross Validation

In [30]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
from sklearn.model_selection import cross_val_score

cross_val_score(LinRegression, X_test, y_test, cv=5)

array([0.89691674, 0.81902374, 0.8401479 , 0.80107691, 0.84473187])

In [18]:
# Report on their finding and their significance

All five integers are really close to eachother and are trying to get close to the true value of LineRegression.

## Part 5. Using Polynomial Regression

In [28]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
from sklearn.preprocessing import PolynomialFeatures
from sklearn import linear_model

PolyRegression = PolynomialFeatures(degree=2)
PolyRegression_train = PolyRegression.fit_transform(X_train)
PolyRegression_test = PolyRegression.fit_transform(X_test)

BayRegression = linear_model.BayesianRidge().fit(PolyRegression_train, y_train)
BayRegression.score(PolyRegression_test, y_test)

1.0

This means 100% accuracy.

In [29]:
# Report on the metrics and output the resultant equation as you did in Part 3.
print("Coefficients: ",BayRegression.coef_)
print("Intercept: ",BayRegression.intercept_)
print("Prediction: ", BayRegression.predict(np.array([[1, 2, 3, 4, 5, 6]])))

Coefficients:  [ 0.00000000e+00  1.20000000e+01 -9.44954524e-08  1.36796451e-11
  2.00000000e+00  2.85714287e-02]
Intercept:  1.433130819350481e-05
Prediction:  [34.17144259]


$y = 1.42215(10^{-5}) + 0(1) + 1.2(10^2)(a) - 8.31525e(10^{-8})(b) - 5.34731(10^
{-12})(a^2) + 2(ab) + 2.85714(10^{-2})(b^2)$
<p>We know that the accuracy of this model is 100% which means polynomial regression is much more accurate than the linear regression.</p>
